### PR TITLE
Upgrade react-native to 0.64.2

### PR DESCRIPTION
**Pull request by laurentS**
This PR is fairly messy and is a continuation of #519 (which should be merged before this one).

While moving to github actions, we ended up hitting a problem with uploading to testflight (see [my other comment](https://github.com/mapswipe/mapswipe/pull/519#issuecomment-860061320)) which more or less forced us to upgrade:
- react-native to 0.64.2 (the latest at the time)
- react-native-firebase to v6+ as the previous versions were not supported anymore
- a few other dependencies

This PR tracks these upgrades, and multiple attempts to fix a crash on app launch on iOS which was due to an environment variable missing in CI for iOS :roll_eyes: 

Issues covered by this PR:
- #405
- #216 
- #502

### CODE REVIEW

File: .flowconfig [hunk 1]
```diff
@@ -37,24 +34,22 @@ emoji=true
 esproposal.optional_chaining=enable
 esproposal.nullish_coalescing=enable
 
+exact_by_default=true
```
Comment (Hagellach37) [inline .flowconfig:37]: that's how we do it. by default, also exact. :D

but for curiosity: what does it mean?

File: .github/workflows/android.yml [hunk 2]
```diff
@@ -12,26 +12,33 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v1
+
       - uses: actions/setup-node@v1
         with:
-          node-version: '16.1.0'
+          node-version: '14.16.1'
```
Comment (Hagellach37) [inline .github/workflows/android.yml:18]: could we here move again up to 16.1.0? or are there any other reason to use v14?

File: .github/workflows/ios.yml [hunk 3]
```diff
@@ -12,69 +12,129 @@ jobs:
...
     steps:
       - uses: actions/checkout@v1
+
+      - name: List available versions of Xcode
+        run: ls -n /Applications/ | grep Xcode*
+
       - uses: actions/setup-node@v1
         with:
-          node-version: '16.1.0'
+          node-version: '14.16.1'
```
Comment (Hagellach37) [inline .github/workflows/ios.yml:21]: same question as above

File: .github/workflows/ios.yml [hunk 4]
```diff
@@ -12,69 +12,129 @@ jobs:
...
         with:
           ruby-version: '2.7.3'
+
       - name: Install Fastlane
         run: bundle install
+
       - name: Install cocoapods
         run: gem install xcpretty cocoapods
+
+      - name: TEMP manually run pod install
```
Comment (Hagellach37) [inline .github/workflows/ios.yml:36]: Does `TEMP` indicate here that you actually planned to remove these steps?

File: .github/workflows/ios.yml [hunk 5]
```diff
@@ -12,69 +12,129 @@ jobs:
...
           MATCH_PASSWORD: ${{ secrets.MATCH_PASSWORD }}
         run: cd ios && bundle exec fastlane release
+
+      - name: Upload build artifacts
+        uses: actions/upload-artifact@v2
+        with:
+          name: build.log
+          path: ~/Library/Logs/gym/*.log
+
+      - name: Find output files
```
Comment (Hagellach37) [inline .github/workflows/ios.yml:128]: do we still need this? looks like you tried to log which files there are

File: android/app/src/main/AndroidManifest.xml [hunk 6]
```diff
@@ -16,22 +16,14 @@
...
+        android:configChanges="keyboard|keyboardHidden|orientation|screenSize|uiMode"
+        android:launchMode="singleTask"
         android:windowSoftInputMode="adjustResize">
         <intent-filter>
             <action android:name="android.intent.action.MAIN" />
             <category android:name="android.intent.category.LAUNCHER" />
         </intent-filter>
       </activity>
-      <activity android:name="com.facebook.react.devsupport.DevSettingsActivity" />
-      <service android:name="io.invertase.firebase.messaging.RNFirebaseMessagingService">
```
Comment (Hagellach37) [inline android/app/src/main/AndroidManifest.xml:29]: I think we've said this somewhere else already. but this means that we can't send notifications, right?

File: android/app/src/main/java/org/missingmaps/mapswipe/MainActivity.java [hunk 7]
```diff
@@ -5,7 +5,6 @@
 import com.facebook.react.ReactActivityDelegate;
 import com.facebook.react.ReactRootView;
 import com.swmansion.gesturehandler.react.RNGestureHandlerEnabledRootView;
-import org.devio.rn.splashscreen.SplashScreen;
```
Comment (Hagellach37) [inline android/app/src/main/java/org/missingmaps/mapswipe/MainActivity.java:8]: do we want to have the splashscreen again? Or is there any other reason, why we should remove it?

File: android/app/src/main/java/org/missingmaps/mapswipe/MainApplication.java [hunk 8]
```diff
@@ -6,20 +6,10 @@
 import com.facebook.react.PackageList;
 
 import com.facebook.react.ReactApplication;
-import io.invertase.firebase.RNFirebasePackage;
+import com.facebook.react.ReactInstanceManager;
 import com.BV.LinearGradient.LinearGradientPackage;
 import com.reactnativecommunity.asyncstorage.AsyncStoragePackage;
-import io.sentry.RNSentryPackage;
```
Comment (Hagellach37) [inline android/app/src/main/java/org/missingmaps/mapswipe/MainApplication.java:12]: did you remove sentry as well? or do we still get the errors reports in sentry?

File: android/app/src/main/java/org/missingmaps/mapswipe/MainApplication.java [hunk 9]
```diff
@@ -6,20 +6,10 @@
...
-import com.reactnativecommunity.viewpager.RNCViewPagerPackage;
 import com.reactnativecommunity.netinfo.NetInfoPackage;
-import io.invertase.firebase.RNFirebasePackage;
```
Comment (Hagellach37) [inline android/app/src/main/java/org/missingmaps/mapswipe/MainApplication.java:15]: all these packages have been removed due to the upgrade of the firebase package? and are now covered by something else?

File: ios/Podfile [hunk 10]
```diff
@@ -1,72 +1,36 @@
...
 
   target 'mapswipeTests' do
-    inherit! :search_paths
+    inherit! :complete
     # Pods for testing
   end
 
-  use_native_modules!
-end
+  # Enables Flipper.
```
Comment (Hagellach37) [inline ios/Podfile:26]: becomes it has come up so often: What is the flipper?

File: src/shared/views/Welcome.js [hunk 11]
```diff
@@ -68,14 +67,14 @@ class _WelcomeScreen extends React.Component<Props> {
         if (welcomeCompleted) {
             this.finishWelcomeScreens();
         } else {
-            SplashScreen.hide();
+            // SplashScreen.hide();
```
Comment (Hagellach37) [inline src/shared/views/Welcome.js:70]: again here: do we want the splashscreen or not?

**Comment by Hagellach37** [comment]
Hey @laurentS ,
most things are clear. Just some questions above on node version and if we want to have the splashscreen or not.

File: .flowconfig (hunk 1 above)
Comment (laurentS) [reply .flowconfig:37]: Honestly, I'm not entirely sure, I just followed the upgrade docs. flow was not on a priority list as it's never blocking releases anyway. I've put a lot of `$FlowFixMe` comments in the code just to get CI to pass, I suggest we leave resolving these for a separate PR, once we're back on track.

File: .github/workflows/android.yml (hunk 2 above)
Comment (laurentS) [reply .github/workflows/android.yml:18]: Indeed, I think we can upgrade again, the downgrade was just to test if this was causing crashes. Done in 3b5bf5bd

File: .github/workflows/ios.yml (hunk 3 above)
Comment (laurentS) [reply .github/workflows/ios.yml:21]: Same answer as above :)

File: .github/workflows/ios.yml (hunk 4 above)
Comment (laurentS) [reply .github/workflows/ios.yml:36]: yes, done!

File: .github/workflows/ios.yml (hunk 5 above)
Comment (laurentS) [reply .github/workflows/ios.yml:128]: Removed, good catch.

File: android/app/src/main/AndroidManifest.xml (hunk 6 above)
Comment (laurentS) [reply android/app/src/main/AndroidManifest.xml:29]: Yes, this boils down to RNfirebase having to be upgraded to v6, in which notification management has been extracted to a separate component. I will have to put them back in, for the third time :(

File: android/app/src/main/java/org/missingmaps/mapswipe/MainActivity.java (hunk 7 above)
Comment (laurentS) [reply android/app/src/main/java/org/missingmaps/mapswipe/MainActivity.java:8]: Two different things at play here:
- RN's import system works slightly differently now, so this line is not needed
- the splashscreen package we were using is not supported anymore, so I've removed it (first to check if it was causing the crash, but also to replace it with a properly maintained one, which I'll do asap)

File: android/app/src/main/java/org/missingmaps/mapswipe/MainApplication.java (hunk 8 above)
Comment (laurentS) [reply android/app/src/main/java/org/missingmaps/mapswipe/MainApplication.java:12]: See point above: the import system is slightly different in this version of RN, we still very much want sentry to report errors, as we'd be virtually blind without it.

File: android/app/src/main/java/org/missingmaps/mapswipe/MainApplication.java (hunk 9 above)
Comment (laurentS) [reply android/app/src/main/java/org/missingmaps/mapswipe/MainApplication.java:15]: Yes, see previous comments, except for notifications, everything else is still there, but these lines are not needed anymore.

File: ios/Podfile (hunk 10 above)
Comment (laurentS) [reply ios/Podfile:26]: Flipper is RN's new debugger (for javascript, I think). I've not even had a chance to look at it yet, but apparently it's good.

File: src/shared/views/Welcome.js (hunk 11 above)
Comment (laurentS) [reply src/shared/views/Welcome.js:70]: It's coming back very soon, but I think I'll open a separate PR for it, just so we can merge this and get back to normal life.

**Comment by laurentS** [comment]
Comments addressed, and changes pushed back, if CI passes and you have no other comments, I think we're good to merge.

**Comment by Hagellach37** [comment]
Hey,
yep let's merge this one.

**Review by Hagellach37 — APPROVED** [review]

**Pull request merged**